Add quantization noise + remove sky_image quantization

aguinot opened the pull request:
Add quantization noise to the image. This allow to improve background estimation, see #85.
Below are the results with the quantization noise on. It looks like the original signal is not impacted.

![backgrounds](https://github.com/user-attachments/assets/3ae399d1-547c-4c22-901d-6fd756b2b252)

![background_hist](https://github.com/user-attachments/assets/fbced1fe-b8c1-4de4-be39-f65cc2906fd7)

aguinot commented:
Question:
At the moment the quantization noise is always added as long as any is requested. All the other noise have a dedicated keyword in the config to turn on/off. Do we want that for this too? (very easy to add)

FedericoBerlfein commented on euclidlike_imsim/noise.py:
```diff
@@ -97,8 +97,15 @@ def get_noise(cfg_noise, cfg_image, base, logger):
+    quantization_noise.fill(0)
+    quantization_noise.addNoise(galsim.DeviateNoise(galsim.UniformDeviate(rng)))
+    noise_img += quantization_noise
+    noise_img -= 0.5
```
Where does this subtraction by 0.5 come from?

aguinot replied:
With quantization we assume that if you have a value of 100.5 ADUs it has the same chance to round to 100 or 101 which lead to +/-0.5. The Uniform distribution from GalSim return value in the range [0, 1], the -0.5 is just to rescale the range to [-0.5, 0.5].

rmandelb commented on euclidlike_imsim/noise.py:
```diff
@@ -97,8 +97,15 @@ def get_noise(cfg_noise, cfg_image, base, logger):
     noise_img.quantize()
 
+    # Add quantization noise (this avoid issues related to rounding e.g.
+    # backgroud estimation)
```
```suggestion
    # background estimation)
```

rmandelb replied:
Also, it could be good to give a link to the ERO paper and mention the section that discusses this, and maybe give a simple high-level summary, like "this makes it easier to process the images, without modifying the expectation value of the signal"

aguinot commented:
> As discussed during the meeting:
> 
> * put in documentation
> * include in end to end demo

- I have added a basic documentation to the `get_noise` function which will be propagated to the documentation (it i the easiest as there is no dedicated page in addition to the API).
- Added to the end-to-end demo

rmandelb commented on euclidlike_imsim/noise.py:
```diff
@@ -39,6 +40,19 @@ def parse_noise_config(params):
+    Parameters:
+        cfg_noise (dict): Configuration dictionary for noise parameters.
+
+    NOTE on the quantization: This prevent to have problems due to the rounding.
```
```suggestion
    NOTE on the quantization: This prevents problems due to the rounding.
```

rmandelb commented on euclidlike_imsim/noise.py:
```diff
@@ -39,6 +40,19 @@ def parse_noise_config(params):
+    NOTE on the quantization: This prevent to have problems due to the rounding.
+    For example, given the very low amplitude of the sky background, we can have
+    spatial variations of only ~1 ADU. This will be impossible to properly pick
+    up by tools SExtractor. Adding this noise prvent this problem and do not
```
```suggestion
    up by tools like SExtractor. Adding this noise prevents this problem and does not
```

rmandelb commented on euclidlike_imsim/noise.py:
```diff
@@ -39,6 +40,19 @@ def parse_noise_config(params):
+    For example, given the very low amplitude of the sky background, we can have
+    spatial variations of only ~1 ADU. This will be impossible to properly pick
+    up by tools SExtractor. Adding this noise prvent this problem and do not
+    change the signal in the image. This is discuss in [Cuillandre et al. 2025](https://arxiv.org/abs/2405.13496)
```
```suggestion
    change the signal in the image. This is discussed in [Cuillandre et al. 2025](https://arxiv.org/abs/2405.13496)
```

rmandelb commented on euclidlike_imsim/noise.py:
```diff
@@ -97,8 +111,16 @@ def get_noise(cfg_noise, cfg_image, base, logger):
+    if cfg_noise["quantization_noise"]:
+        quantization_noise = noise_img.copy()
+        quantization_noise.fill(0)
+        quantization_noise.addNoise(galsim.DeviateNoise(galsim.UniformDeviate(rng)))
```
On the github interface something funny is happening with indentation?  Is this a tabs vs. spaces thing or just a fluke?

rmandelb commented on examples/end_to_end_demo.py:
```diff
@@ -395,10 +395,21 @@ def main(argv):
         # Finally, the analog-to-digital converter reads in an integer value.
         full_image.quantize()
-        sky_image.quantize()
+        # sky_image.quantize() # Quantizinf the background will actually lead to some issues.
```
```suggestion
        # sky_image.quantize() # Quantizing the background will actually lead to some issues.
```

rmandelb commented on examples/end_to_end_demo.py:
```diff
@@ -395,10 +395,21 @@ def main(argv):
         # Note that the image type after this step is still a float.  If we want to actually
         # get integer values, we can do new_img = galsim.Image(full_image, dtype=int)
 
+        # Here we add quantization noise to the image. This prevent to have problems due
```
```suggestion
        # Here we add quantization noise to the image. This prevents problems due
```

rmandelb commented on examples/end_to_end_demo.py:
```diff
@@ -395,10 +395,21 @@ def main(argv):
+        # Here we add quantization noise to the image. This prevent to have problems due
+        # to the rounding. For example, given the very low amplitude of the sky background,
+        # we can have spatial variations of only ~1 ADU. This will be impossible to 
+        # properly pick up by tools SExtractor. Adding this noise prvent this problem
```
```suggestion
        # properly pick up by tools like SExtractor. Adding this noise prevents this problem
```

rmandelb commented on examples/end_to_end_demo.py:
```diff
@@ -395,10 +395,21 @@ def main(argv):
+        # to the rounding. For example, given the very low amplitude of the sky background,
+        # we can have spatial variations of only ~1 ADU. This will be impossible to 
+        # properly pick up by tools SExtractor. Adding this noise prvent this problem
+        # and do not change the signal in the image.
```
```suggestion
        # and does not change the signal in the image.
```

rmandelb reviewed (approved):
All my comments/suggestions are trivial, so I'm approving to avoid the need for another review.  I think you should be able to just accept the suggestions and merge.
